clients/src/main/java/org/apache/kafka/clients/consumer/PartitionAssignor.java

```diff
@@ -0,0 +1,125 @@
+     * @param consumerId The identifier of the local consumer instance
+     * @param consumers Map of all consumers in the group and their respective metadata
+     * @param metadataSnapshot The metadata snapshot that was used when the group was created, including
+     *                         subscription information and topic metadata
+     * @return
+     */
+    AssignmentResult assign(String consumerId,
+                            Map<String, M> consumers,
+                            MetadataSnapshot metadataSnapshot);
+
+    /**
+     * The name of the partition assignor (e.g. 'roundrobin')
+     * @return The name (must not be null)
+     */
+    String name();
+
+    /**
+     * The version of the partition assignor
+     * @return The version
+     */
+    short version();
+
+    /**
+     * Schema representing this assignor's metadata.
+     * @return Non-null schema
+     */
+    Type schema();
+
+    /**
+     * Get the local metadata for this assignor. This will be invoked prior to every
+     * round of group membership, so it is allowed to change dynamically. The metadata
+     * will probably include at least the subscription of the consumer, but can also include
+     * local information such as the number of processors on the host or its fqdn. Note,
+     * however, that extra care must be taken when designing an assignor for large groups
+     * since the metadata from each member must be propagated to ALL members of the group.
+     *
+     * @param metadataSnapshot Snapshot of the consumer's metadata/subscription.
+     * @return The latest metadata for
+     */
+    M metadata(MetadataSnapshot metadataSnapshot);
+
+    /**
+     * Wrapper
+     */
+    class AssignmentResult {
+        private boolean succeeded = false;
+        private List<TopicPartition> assignment;
+        private Set<String> groupSubscription;
+
+        public AssignmentResult(boolean succeeded,
+                                List<TopicPartition> assignment,
+                                Set<String> groupSubscription) {
+            this.succeeded = succeeded;
+            this.assignment = assignment;
+            this.groupSubscription = groupSubscription;
+        }
+
+        public boolean succeeded() {
+            return succeeded;
+        }
+
+        public List<TopicPartition> assignment() {
+            return assignment;
+        }
+
+        public Set<String> groupSubscription() {
+            return groupSubscription;
+        }
+
+        /**
+         * Successful assignment.
+         * @param assignment The partitions assigned to the local group member
+         * @return AssignmentResult indicated the success
+         */
+        public static AssignmentResult success(List<TopicPartition> assignment) {
+            return new AssignmentResult(true, assignment, null);
+        }
+
+        /**
+         * Failed assignment. Assignments can fail from inconsistent metadata. In that case, we need
```

Comment:
Can we may be document what `groupSubscription` is more clearly somewhere else, maybe even just on this class's javadoc? It took me a bit of digging to figure out what it was -- I was confused why we were getting (and using) something called group subscription when a join group had failed.

Maybe a bit of renaming could also help? aggregateGroupSubscription or aggregateRequestedGroupSubscription?